optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/VehicleRoutingConstraintProvider.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaweb.vehiclerouting.plugin.planner;
+
+import org.optaplanner.core.api.score.buildin.hardsoftlong.HardSoftLongScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintCollectors;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaweb.vehiclerouting.plugin.planner.domain.PlanningVisit;
+
+public class VehicleRoutingConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[]{
+                vehicleCapacityExceeded(constraintFactory),
+                distanceFromPreviousStandstill(constraintFactory),
+                distanceFromLastVisitToDepot(constraintFactory)
+        };
+    }
+
+    private static Constraint vehicleCapacityExceeded(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(PlanningVisit.class)
+                .groupBy(PlanningVisit::getVehicle, ConstraintCollectors.sum(PlanningVisit::getDemand))
+                .filter((vehicle, demand) -> demand > vehicle.getCapacity())
+                .penalizeLong(
+                        "vehicle capacity exceeded",
+                        HardSoftLongScore.ONE_HARD,
+                        (vehicle, demand) -> demand - vehicle.getCapacity()
+                );
+    }
+
+    private static Constraint distanceFromPreviousStandstill(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(PlanningVisit.class)
+                .filter(visit -> visit.getPreviousStandstill() != null)
```

Comment:
Personally, I'd replace the filter with `ifExists(..., Joiner.equals(...))`.
Filters are not indexed, leading to a little drop in performance.
More importantly, I consider that easier to understand than the `null` check.

Reply:
Turns out the filter is not necessary here.